File: algorithms/algorithms-cplusplus/leetcode/Question_0061_Rotate_List.cc

//Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* rotateRight(ListNode* head, int k) {
    if (!head || k <= 0) {
      return head;
    }
    int n = 1;
    ListNode *cur = head;
    while (cur->next) {
      cur = cur->next;
      n++;
    }
    cur->next = head;
    n -= k % n;
    while (n--) {
      cur = cur->next;
    }
    head = cur->next;
    cur->next = nullptr;
    return head;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0061_Rotate_List.py
#!/usr/bin/python
# coding: utf-8

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or k < 0:
            return None
        if k == 0:
            return head
        n, cur = 1, head
        while cur.next:
            n += 1
            cur = cur.next
        cur.next = head
        n -= k % n
        while n > 0:
            n -= 1
            cur = cur.next
        head = cur.next
        cur.next = None
        return head
